control-plane/piSerializer.cpp

```diff
@@ -0,0 +1,1471 @@
+
+  // Iterate over the fields, recursively handling nested structures if necessary.
+  size_t offset = 0;
+  for (auto field : *structType->fields) {
+    forAllFieldsInDeclaration(path, offset++, field, typeMap, function);
+  }
+
+  if (structType->is<IR::Type_Header>()) {
+    // Synthesize a '_valid' field, which is used to implement `isValid()`.
+    function(HeaderField::synthesize(path, "_valid", offset, /* bitwidth = */ 1));
+  }
+}
+
+template <typename Func> 
+static void forAllFieldsInStack(const HeaderFieldPath* path,
+                                const TypeMap* typeMap,
+                                Func function) {
+  auto stackType = path->type->to<IR::Type_Stack>();
+  auto elementType = typeMap->getTypeType(stackType->elementType, true);
+  BUG_CHECK(elementType->is<IR::Type_Header>(),
+            "Header stack %1% has non-header element type %2%",
+            path->serialize(), elementType);
+
+  for (unsigned stackIndex = 0; stackIndex < stackType->getSize(); stackIndex++) {
+    forAllFieldsInStructlike(path->append(stackIndex, elementType),
+                             typeMap, function);
+  }
+}
+
+template <typename Func>
+static void forAllFieldsInDeclaration(const HeaderFieldPath* path,
+                                      size_t offset,
+                                      const IR::IDeclaration* field,
+                                      const TypeMap* typeMap,
+                                      Func function) {
+  CHECK_NULL(field);
+
+  auto id = externalId(field);
+  auto name = field->externalName();
+  auto annotations = field->to<IR::IAnnotated>();
+
+  auto fieldType = typeMap->getType(field->getNode(), true);
+  if (fieldType->is<IR::Type_Boolean>()) {
+    const uint32_t bitwidth = 1;
+    function(HeaderField::from(path, name, offset, bitwidth, id, annotations));
+  } else if (fieldType->is<IR::Type_Bits>()) {
+    const uint32_t bitwidth = fieldType->to<IR::Type_Bits>()->size;
+    function(HeaderField::from(path, name, offset, bitwidth, id, annotations));
+  } else if (fieldType->is<IR::Type_StructLike>()) {
+    forAllFieldsInStructlike(path->append(name, fieldType), typeMap, function);
+  } else if (fieldType->is<IR::Type_Stack>()) {
+    forAllFieldsInStack(path->append(name, fieldType), typeMap, function);
+  } else {
+    BUG("Header field '%1%' has unhandled type %2%", name, fieldType);
+  }
+}
+
+} // namespace detail
+
+/// Invoke @function for every primitive header field reachable from
+/// @declaration by recursively expanding all composite fields.
+template <typename Func>
+static void forAllHeaderFields(const IR::IDeclaration* declaration,
+                               const TypeMap* typeMap,
+                               Func function) {
+  CHECK_NULL(declaration);
+  CHECK_NULL(typeMap);
+
+  auto type = typeMap->getType(declaration->getNode(), true);
+  if (isMetadataType(type)) {
+    // Unlike other top-level structs, top-level metadata structs are fully
```

Comment:
Neither do I. This is something we should fix; we should expose structs uniformly to PI. What I've done here to document the current approach, which is extremely hacky.

We were talking about this issue elsewhere; see here:

https://github.com/p4lang/p4c/issues/225#issuecomment-272339779